systems/controllers/osc/operational_space_control.cc

```diff
@@ -0,0 +1,602 @@
+      std::string name_w_spr = tree_w_spr_.get_position_name(j);
+      if (name_wo_spr.compare(0, name_wo_spr.size(), name_w_spr) == 0) {
+        map_position_from_spring_to_no_spring_(i, j) = 1;
+      }
+    }
+  map_velocity_from_spring_to_no_spring_ = MatrixXd::Zero(n_v_, n_v_w_spr);
+  for (int i = 0; i < n_v_; i++)
+    for (int j = 0; j < n_v_w_spr; j++) {
+      std::string name_wo_spr = tree_wo_spr_.get_velocity_name(i);
+      std::string name_w_spr = tree_w_spr_.get_velocity_name(j);
+      if (name_wo_spr.compare(0, name_wo_spr.size(), name_w_spr) == 0) {
+        map_velocity_from_spring_to_no_spring_(i, j) = 1;
+      }
+    }
+
+  // Get input limits
+  VectorXd u_min(n_u_);
+  VectorXd u_max(n_u_);
+  for (int i = 0; i < n_u_; i++) {
+    u_min(i) = tree_wo_spr_.actuators[i].effort_limit_min_;
+    u_max(i) = tree_wo_spr_.actuators[i].effort_limit_max_;
+  }
+  u_min_ = u_min;
+  u_max_ = u_max;
+
+  // Check if the model is floating based
+  is_quaternion_ = multibody::IsFloatingBase(tree_w_spr);
+}
+
+// Cost methods
+void OperationalSpaceControl::AddAccelerationCost(int joint_vel_idx, double w) {
+  if (W_joint_accel_.size() == 0) {
+    W_joint_accel_ = Eigen::MatrixXd::Zero(n_v_, n_v_);
+  }
+  W_joint_accel_(joint_vel_idx, joint_vel_idx) += w;
+}
+
+// Constraint methods
+void OperationalSpaceControl::AddContactPoint(int body_index,
+    Eigen::VectorXd pt_on_body) {
+  body_index_.push_back(body_index);
+  pt_on_body_.push_back(pt_on_body);
+}
+void OperationalSpaceControl::AddStateAndContactPoint(int state,
+    int body_index, Eigen::VectorXd pt_on_body) {
+  state_.push_back(state);
+  AddContactPoint(body_index, pt_on_body);
+}
+void OperationalSpaceControl::AddContactPoint(std::vector<int> body_index,
+    std::vector<Eigen::VectorXd> pt_on_body) {
+  body_index_.insert(body_index_.end(), body_index.begin(), body_index.end());
+  pt_on_body_.insert(pt_on_body_.end(), pt_on_body.begin(), pt_on_body.end());
+}
+void OperationalSpaceControl::AddStateAndContactPoint(std::vector<int> state,
+    std::vector<int> body_index,
+    std::vector<Eigen::VectorXd> pt_on_body) {
+  state_.insert(state_.end(), state.begin(), state.end());
+  AddContactPoint(body_index, pt_on_body);
+}
+
+// Osc checkers and constructor
+void OperationalSpaceControl::CheckCostSettings() {
+  if (W_input_.size() != 0) {
+    DRAKE_DEMAND((W_input_.rows() == n_u_) && (W_input_.cols() == n_u_));
+  }
+  if (W_joint_accel_.size() != 0) {
+    DRAKE_DEMAND((W_joint_accel_.rows() == n_v_) &&
+                 (W_joint_accel_.cols() == n_v_));
+  }
+}
+void OperationalSpaceControl::CheckConstraintSettings() {
+  if (!body_index_.empty()) {
+    DRAKE_DEMAND(mu_ != -1);
+    DRAKE_DEMAND(body_index_.size() == pt_on_body_.size());
+  }
+  if (!state_.empty()) {
+    DRAKE_DEMAND(state_.size() == body_index_.size());
+    DRAKE_DEMAND(state_.size() == pt_on_body_.size());
+  }
+}
+
+void OperationalSpaceControl::ConstructOSC() {
```

Comment:
This isn't a constructor, so maybe `Compile` rather than `ConstructOSC`? It should also have a `is_compiled_` flag that's set, with appropriate errors if (1) compile is called twice or (2) it's run without being compiled.

Reply:
I was looking for the verb, but couldn't think of a good one. Do you think _compile_ is fine because it also has the meaning of translating our code to machine code?